App crashes because of malformed background URL

Here is stacktrace received from Play Store.


```
java.lang.RuntimeException: Unable to start activity ComponentInfo{info.hossainkhan.dailynewsheadlines/info.hossainkhan.dailynewsheadlines.details.HeadlinesDetailsActivity}: android.view.InflateException: Binary XML file line #28: Binary XML file line #28: Error inflating class fragment
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2458)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2518)
	at android.app.ActivityThread.-wrap11(ActivityThread.java)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1345)
	at android.os.Handler.dispatchMessage(Handler.java:104)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5482)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
Caused by: android.view.InflateException: Binary XML file line #28: Binary XML file line #28: Error inflating class fragment
	at android.view.LayoutInflater.inflate(LayoutInflater.java:539)
	at android.view.LayoutInflater.inflate(LayoutInflater.java:423)
	at android.view.LayoutInflater.inflate(LayoutInflater.java:374)
	at com.android.internal.policy.PhoneWindow.setContentView(PhoneWindow.java:393)
	at android.app.Activity.setContentView(Activity.java:2166)
	at info.hossainkhan.dailynewsheadlines.details.HeadlinesDetailsActivity.onCreate(Unknown Source)
	at android.app.Activity.performCreate(Activity.java:6237)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2411)
	... 9 more
Caused by: android.view.InflateException: Binary XML file line #28: Error inflating class fragment
	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:782)
	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:704)
	at android.view.LayoutInflater.rInflate(LayoutInflater.java:835)
	at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798)
	at android.view.LayoutInflater.inflate(LayoutInflater.java:515)
	... 17 more
Caused by: java.lang.IllegalArgumentException: Path must not be empty.
	at com.squareup.picasso.Picasso.load(Picasso.java:297)
	at info.hossainkhan.dailynewsheadlines.details.HeadlinesDetailsFragment.showHeadlineDetails(Unknown Source)
	at info.hossainkhan.android.core.headlines.HeadlinesDetailsViewPresenter.initView(HeadlinesDetailsViewPresenter.java:51)
	at info.hossainkhan.android.core.headlines.HeadlinesDetailsViewPresenter.<init>(HeadlinesDetailsViewPresenter.java:46)
	at info.hossainkhan.dailynewsheadlines.details.HeadlinesDetailsFragment.onCreate(Unknown Source)
	at android.app.Fragment.performCreate(Fragment.java:2198)
	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:942)
	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1126)
	at android.app.FragmentManagerImpl.addFragment(FragmentManager.java:1228)
	at android.app.FragmentManagerImpl.onCreateView(FragmentManager.java:2201)
	at android.app.FragmentController.onCreateView(FragmentController.java:98)
	at android.app.Activity.onCreateView(Activity.java:5546)
	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:754)
	... 21 more
```

--- a/core-lib/src/main/java/info/hossainkhan/android/core/headlines/HeadlinesContract.java
+++ b/core-lib/src/main/java/info/hossainkhan/android/core/headlines/HeadlinesContract.java
@@ -45,7 +45,7 @@ public interface HeadlinesContract {
 
         void showAppSettingsScreen();
 
-        void showHeadlineBackdropBackground(URI imageURI);
+        void showHeadlineBackdropBackground(String imageUrl);
 
         /**
          * Toggles loading indicator based on boolean flag.

--- a/core-lib/src/main/java/info/hossainkhan/android/core/headlines/HeadlinesContract.java
+++ b/core-lib/src/main/java/info/hossainkhan/android/core/headlines/HeadlinesContract.java
@@ -26,14 +26,13 @@ package info.hossainkhan.android.core.headlines;
 
 import android.support.annotation.NonNull;
 
-import java.net.URI;
 import java.util.List;
 
 import info.hossainkhan.android.core.base.MvpPresenter;
 import info.hossainkhan.android.core.base.MvpView;
 import info.hossainkhan.android.core.model.CardItem;
-import info.hossainkhan.android.core.model.ScreenType;
 import info.hossainkhan.android.core.model.NavigationRow;
+import info.hossainkhan.android.core.model.ScreenType;
 
 public interface HeadlinesContract {
 

--- a/core-lib/src/main/java/info/hossainkhan/android/core/headlines/HeadlinesDetailsContract.java
+++ b/core-lib/src/main/java/info/hossainkhan/android/core/headlines/HeadlinesDetailsContract.java
@@ -40,6 +40,8 @@ public interface HeadlinesDetailsContract {
         void openArticleWebUrl(String contentUrl);
 
         void showHeadlineDetails(CardItem cardItem);
+
+        void loadDetailsImage(String imageUrl);
     }
 
     interface Presenter extends MvpPresenter<HeadlinesDetailsContract.View> {

--- a/core-lib/src/main/java/info/hossainkhan/android/core/headlines/HeadlinesDetailsViewPresenter.java
+++ b/core-lib/src/main/java/info/hossainkhan/android/core/headlines/HeadlinesDetailsViewPresenter.java
@@ -30,6 +30,7 @@ import android.content.Context;
 import info.hossainkhan.android.core.R;
 import info.hossainkhan.android.core.base.BasePresenter;
 import info.hossainkhan.android.core.model.CardItem;
+import info.hossainkhan.android.core.util.StringUtils;
 import timber.log.Timber;
 
 public class HeadlinesDetailsViewPresenter extends BasePresenter<HeadlinesDetailsContract.View>
@@ -49,7 +50,14 @@ public class HeadlinesDetailsViewPresenter extends BasePresenter<HeadlinesDetail
     private void initView() {
         getView().updateScreenTitle(mContext.getString(R.string.detail_view_title));
         getView().showHeadlineDetails(mCardItem);
+
+        String imageUrl = mCardItem.imageUrl();
+        if(StringUtils.isValidUri(imageUrl)) {
+            getView().loadDetailsImage(imageUrl);
+        }
     }
+
+
 
 
     @Override

--- a/core-lib/src/main/java/info/hossainkhan/android/core/headlines/HeadlinesPresenter.java
+++ b/core-lib/src/main/java/info/hossainkhan/android/core/headlines/HeadlinesPresenter.java
@@ -39,6 +39,7 @@ import info.hossainkhan.android.core.model.CardItem;
 import info.hossainkhan.android.core.model.ScreenType;
 import info.hossainkhan.android.core.model.NavigationRow;
 import info.hossainkhan.android.core.newsprovider.NewsProviderManager;
+import info.hossainkhan.android.core.util.StringUtils;
 import rx.Observable;
 import rx.Subscriber;
 import rx.Subscription;
@@ -103,8 +104,9 @@ public class HeadlinesPresenter extends BasePresenter<HeadlinesContract.View> im
     @Override
     public void onHeadlineItemSelected(@NonNull final CardItem cardItem) {
         CoreApplication.getAnalyticsReporter().reportHeadlineSelectedEvent(cardItem);
-        if (cardItem.imageUrl() !=null) {
-            getView().showHeadlineBackdropBackground(cardItem.getImageURI());
+        String imageUrl = cardItem.imageUrl();
+        if (StringUtils.isValidUri(imageUrl)) {
+            getView().showHeadlineBackdropBackground(imageUrl);
         } else {
             Timber.i("Card object does not have HD background.");
         }

--- a/core-lib/src/main/java/info/hossainkhan/android/core/model/CardItem.java
+++ b/core-lib/src/main/java/info/hossainkhan/android/core/model/CardItem.java
@@ -31,8 +31,6 @@ import android.support.annotation.Nullable;
 import com.google.auto.value.AutoValue;
 import com.google.gson.annotations.SerializedName;
 
-import java.net.URI;
-import java.net.URISyntaxException;
 import java.util.List;
 
 import info.hossainkhan.android.core.gson.AutoGson;
@@ -177,18 +175,4 @@ public abstract class CardItem {
                 height);
 
     }
-
-
-
-    public URI getImageURI() {
-        if (imageUrl() == null) return null;
-        try {
-            return new URI(imageUrl());
-        } catch (URISyntaxException e) {
-            Timber.e(e);
-            Timber.w("URI exception: ", imageUrl());
-            return null;
-        }
-    }
-
 }

--- a/core-lib/src/main/java/info/hossainkhan/android/core/util/StringUtils.java
+++ b/core-lib/src/main/java/info/hossainkhan/android/core/util/StringUtils.java
@@ -24,20 +24,34 @@
 
 package info.hossainkhan.android.core.util;
 
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import timber.log.Timber;
+
 public class StringUtils {
 
     public static boolean isNotEmpty(String s) {
-        return s != null && s.isEmpty() == false;
+        return s != null && !s.isEmpty();
     }
 
     public static boolean isEmpty(String s) {
-        return isNotEmpty(s) == false;
+        return !isNotEmpty(s);
     }
 
-    public static boolean isEquals(String a, String b) {
-        if(a == null || b == null) {
-            return false;
+    public static boolean isValidUri(String s) {
+        if (isEmpty(s)) return false;
+
+        boolean isValid;
+
+        try {
+            new URI(s);
+            isValid = true;
+        } catch (URISyntaxException e) {
+            Timber.e(e);
+            Timber.w("URI exception for: %s", s);
+            isValid = false;
         }
-        return a.equals(b);
+        return isValid;
     }
 }

--- a/mobile/src/main/java/info/hossainkhan/dailynewsheadlines/MainActivity.java
+++ b/mobile/src/main/java/info/hossainkhan/dailynewsheadlines/MainActivity.java
@@ -5,7 +5,6 @@ import android.os.Bundle;
 import android.widget.TextView;
 import android.widget.Toast;
 
-import java.net.URI;
 import java.util.List;
 
 import butterknife.BindView;
@@ -14,8 +13,8 @@ import info.hossainkhan.android.core.base.BaseActivity;
 import info.hossainkhan.android.core.headlines.HeadlinesContract;
 import info.hossainkhan.android.core.headlines.HeadlinesPresenter;
 import info.hossainkhan.android.core.model.CardItem;
-import info.hossainkhan.android.core.model.ScreenType;
 import info.hossainkhan.android.core.model.NavigationRow;
+import info.hossainkhan.android.core.model.ScreenType;
 import info.hossainkhan.android.core.newsprovider.NewsProviderManager;
 import timber.log.Timber;
 

--- a/mobile/src/main/java/info/hossainkhan/dailynewsheadlines/MainActivity.java
+++ b/mobile/src/main/java/info/hossainkhan/dailynewsheadlines/MainActivity.java
@@ -96,7 +96,7 @@ public class MainActivity extends BaseActivity implements HeadlinesContract.View
     }
 
     @Override
-    public void showHeadlineBackdropBackground(final URI imageURI) {
-        Timber.d("showHeadlineBackdropBackground() called with: imageURI = [" + imageURI + "]");
+    public void showHeadlineBackdropBackground(final String imageUrl) {
+        Timber.d("showHeadlineBackdropBackground() called with: imageUrl = [" + imageUrl + "]");
     }
 }

--- a/tv/src/main/java/info/hossainkhan/dailynewsheadlines/browser/HeadlinesBrowseFragment.java
+++ b/tv/src/main/java/info/hossainkhan/dailynewsheadlines/browser/HeadlinesBrowseFragment.java
@@ -33,7 +33,6 @@ import android.os.Bundle;
 import android.support.v17.leanback.app.BrowseFragment;
 import android.support.v17.leanback.widget.ArrayObjectAdapter;
 
-import java.net.URI;
 import java.util.List;
 
 import info.hossainkhan.android.core.CoreApplication;

--- a/tv/src/main/java/info/hossainkhan/dailynewsheadlines/browser/HeadlinesBrowseFragment.java
+++ b/tv/src/main/java/info/hossainkhan/dailynewsheadlines/browser/HeadlinesBrowseFragment.java
@@ -186,9 +186,9 @@ public class HeadlinesBrowseFragment extends BrowseFragment implements Headlines
     }
 
     @Override
-    public void showHeadlineBackdropBackground(final URI imageURI) {
-        Timber.d("Loading HD background URL: %s", imageURI);
-        mPicassoBackgroundManager.updateBackgroundWithDelay(imageURI);
+    public void showHeadlineBackdropBackground(final String imageUrl) {
+        Timber.d("Loading HD background URL: %s", imageUrl);
+        mPicassoBackgroundManager.updateBackgroundWithDelay(imageUrl);
     }
 
 

--- a/tv/src/main/java/info/hossainkhan/dailynewsheadlines/details/HeadlinesDetailsFragment.java
+++ b/tv/src/main/java/info/hossainkhan/dailynewsheadlines/details/HeadlinesDetailsFragment.java
@@ -69,6 +69,7 @@ public class HeadlinesDetailsFragment extends DetailsFragment implements Headlin
     private PicassoBackgroundManager mPicassoBackgroundManager;
     private PicassoImageTargetDetailsOverview mDetailsRowPicassoTarget;
     private HeadlinesDetailsViewPresenter mPresenter;
+    private DetailsOverviewRow mDetailsOverview;
 
 
     @Override
@@ -154,26 +155,30 @@ public class HeadlinesDetailsFragment extends DetailsFragment implements Headlin
         mRowsAdapter = new ArrayObjectAdapter(rowPresenterSelector);
 
         // Setup action and detail row.
-        DetailsOverviewRow detailsOverview = new DetailsOverviewRow(cardItem);
+        mDetailsOverview = new DetailsOverviewRow(cardItem);
         // DEV NOTE: Without the image drawable, the details view occupies full width for texts.
-        detailsOverview.setImageDrawable(getResources().getDrawable(R.drawable.placeholder_loading_image));
+        mDetailsOverview.setImageDrawable(getResources().getDrawable(R.drawable.placeholder_loading_image));
 
-        mDetailsRowPicassoTarget = new PicassoImageTargetDetailsOverview(mApplicationContext, detailsOverview);
-        Picasso.with(mApplicationContext)
-                .load(cardItem.imageUrl())
-                .into(mDetailsRowPicassoTarget);
-
-
-        mRowsAdapter.add(detailsOverview);
-
-
+        mRowsAdapter.add(mDetailsOverview);
         setAdapter(mRowsAdapter);
-        mPicassoBackgroundManager.updateBackgroundWithDelay(cardItem.getImageURI(), PicassoBackgroundManager.TransformType.GREYSCALE);
+
 
         // NOTE: Move this when data is loaded
         startEntranceTransition();
     }
 
+    @Override
+    public void loadDetailsImage(final String imageUrl) {
+        mDetailsRowPicassoTarget = new PicassoImageTargetDetailsOverview(mApplicationContext, mDetailsOverview);
+        Picasso.with(mApplicationContext)
+                .load(imageUrl)
+                .into(mDetailsRowPicassoTarget);
+
+        mPicassoBackgroundManager.updateBackgroundWithDelay(
+                imageUrl,
+                PicassoBackgroundManager.TransformType.GREYSCALE);
+
+    }
 
 
     @Override

--- a/tv/src/main/java/info/hossainkhan/dailynewsheadlines/utils/PicassoBackgroundManager.java
+++ b/tv/src/main/java/info/hossainkhan/dailynewsheadlines/utils/PicassoBackgroundManager.java
@@ -34,8 +34,6 @@ import android.util.DisplayMetrics;
 import com.squareup.picasso.Picasso;
 import com.squareup.picasso.RequestCreator;
 
-import java.net.URI;
-import java.net.URISyntaxException;
 import java.util.Timer;
 import java.util.TimerTask;
 
@@ -59,7 +57,7 @@ public class PicassoBackgroundManager {
     private Activity mActivity;
     private BackgroundManager mBackgroundManager = null;
     private DisplayMetrics mMetrics;
-    private URI mBackgroundURI;
+    private String mBackgroundUrl;
     private PicassoImageTarget mBackgroundTarget;
 
     private Timer mBackgroundTimer; // null when no UpdateBackgroundTask is running.
@@ -102,8 +100,8 @@ public class PicassoBackgroundManager {
                 @Override
                 public void run() {
                      /* Here is main (UI) thread */
-                    if (mBackgroundURI != null && StringUtils.isNotEmpty(mBackgroundURI.toString())) {
-                        updateBackground(mBackgroundURI);
+                    if (mBackgroundUrl != null && StringUtils.isNotEmpty(mBackgroundUrl.toString())) {
+                        updateBackground(mBackgroundUrl);
                     } else {
                         updateBackground();
                     }
@@ -116,41 +114,31 @@ public class PicassoBackgroundManager {
      * update backgroud to default
      */
     public void updateBackgroundWithDelay() {
-        mBackgroundURI = null;
+        mBackgroundUrl = null;
         startBackgroundTimer();
     }
 
-
-    public void updateBackgroundWithDelay(String url) {
-        try {
-            URI uri = new URI(url);
-            updateBackgroundWithDelay(uri);
-        } catch (URISyntaxException e) {
-            /* skip updating background */
-            Timber.e(e);
-        }
-    }
 
     /**
      * updateBackground with delay
      * delay time is measured in other Timer task thread.
-     * @param uri Image URI to load
+     * @param imageUrl Image URI to load
      */
-    public void updateBackgroundWithDelay(URI uri) {
-        updateBackgroundWithDelay(uri, null);
+    public void updateBackgroundWithDelay(String imageUrl) {
+        updateBackgroundWithDelay(imageUrl, null);
     }
 
-    public void updateBackgroundWithDelay(URI uri, TransformType transformType) {
-        mBackgroundURI = uri;
+    public void updateBackgroundWithDelay(String imageUrl, TransformType transformType) {
+        mBackgroundUrl = imageUrl;
         mTransformType = transformType;
         startBackgroundTimer();
     }
 
-    private void updateBackground(URI uri) {
+    private void updateBackground(String imageUrl) {
         try {
             Picasso picasso = Picasso.with(mActivity);
             RequestCreator requestCreator = picasso
-                    .load(uri.toString())
+                    .load(imageUrl)
                     .resize(mMetrics.widthPixels, mMetrics.heightPixels)
                     .centerCrop()
                     .error(mDefaultBackground);

